src/types/mod.rs

```diff
@@ -424,6 +426,128 @@ impl ValueInternal for JsNumber {
     }
 }
 
+/// A JavaScript Date object.
+#[repr(C)]
+#[derive(Clone, Copy, Debug)]
+#[cfg(feature = "napi-5")]
+pub struct JsDate(raw::Local);
+
+#[cfg(feature = "napi-5")]
+impl Value for JsDate { }
+
+#[cfg(feature = "napi-5")]
+impl Managed for JsDate {
+    fn to_raw(self) -> raw::Local { self.0 }
+
+    fn from_raw(_: Env, h: raw::Local) -> Self { JsDate(h) }
+}
+
+#[cfg(feature = "napi-5")]
+#[derive(PartialEq, PartialOrd, Clone, Debug)]
+pub struct DateError(f64, DateErrorKind);
+
+#[cfg(feature = "napi-5")]
+impl DateError {
+    pub fn kind(&self) -> DateErrorKind {
+        self.1
+    }
+}
+
+#[cfg(feature = "napi-5")]
+impl fmt::Display for DateError {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        write!(fmt, "{}", self.1.as_str())
+    }
+}
+
+#[cfg(feature = "napi-5")]
+impl Error for DateError {}
+
+#[cfg(feature = "napi-5")]
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Copy, Debug)]
+pub enum DateErrorKind {
+    Overflow,
+    Underflow,
+}
+
+#[cfg(feature = "napi-5")]
+impl DateErrorKind {
+    pub fn as_str(&self) -> &'static str {
+        match *self {
+            DateErrorKind::Overflow => "date overflow",
+            DateErrorKind::Underflow => "date underflow",
+        }
+    }
+}
+
+#[cfg(feature = "napi-5")]
+pub type DateResult<'a> = Result<Handle<'a, JsDate>, DateError>;
+
+#[cfg(feature = "napi-5")]
+impl<'a> JsResultExt<'a, JsDate> for DateResult<'a> {
+    /// Creates a `Error` on error
+    fn or_throw<'b, C: Context<'b>>(self, cx: &mut C) -> JsResult<'a, JsDate> {
+        match self {
+            Ok(v) => Ok(v),
+            Err(e) => cx.throw_error(&e.to_string())
+        }
+    }
+}
+
+#[cfg(feature = "napi-5")]
+impl JsDate {
+    pub const MIN_VALUE: f64 = -8.64e15;
+    pub const MAX_VALUE: f64 = 8.64e15;
+
+    pub fn new<'a, C: Context<'a>, T: Into<f64>>(cx: &mut C, time: T) -> Handle<'a, JsDate> {
```

Comment:
The names on these functions are reversed. `new` should return a `Result` and `new_lossy` should always return the date (which could lose the `f64`).